### ISSUE TITLE
Hi from Beaker!

### ISSUE THREAD

**Issue by pfrazee** [runvnc/noscriptweb#1]
I got directed here by a friend via your HN comment, just wanted to say hi and share some of the stuff we're trying with Beaker browser. (Big note: I'm not at all looking to discourage you to drive forward with your own work! Do it! We just share a lot of ideas and I thought it'd be neat to connect on them.)

Beaker uses a p2p protocol called dat. We baked in Markdown support. We use the p2p protocol to add writability to the web (https://twitter.com/pfrazee/status/1155221334106017792?s=20). The upcoming version will have a version of p2p search along with some basic social network via a kind of RSS-like protocol of syncing down followed sites (https://unwalled.garden).

Ideas around alternative app-code execution are all super interesting. One angle I'm pursuing right now is a sidebar which can run applications against the current page, which is partly what that linked tweet is about (there's a comments app beside the editor app). For running applications directly on the page, I've been toying with two thoughts- the first is to have the site declare "theme" code which is dynamically injected to decorate the page without modifying the file, and the second is to have the browser inject "viewer" code that the user has installed according to properties of the page. 🤷‍♂ Still figuring out how that can work.

**Comment by pfrazee** [comment]
Another thought: I'm a big +1 to building a new browser with different page-models from scratch instead of importing chromium via electron like we're doing. The browser is a huuuuge runtime and I think a clean approach would be worth digging into.

**Comment by runvnc** [comment]
Thanks for telling me about your project.  It looks pretty amazing. I installed it and looked around.  You have nailed making it easy to pick up and go, better than the other p2p web things I have tried.

The performance is also probably slightly better than the other p2p web stuff, but still honestly not as good as the normal web.  Maybe some type of pre-caching of links could help?  It seems like any pause for dat to load a hash or anything could turn into a big sticking issue with adoption.  Because people are impatient.

This github page is a placeholder with just a README, and mainly I just created it this afternoon because wanted to share the ideas I had after reading Doctorow's article today.  I have a job during the week and another side project (3D libretro front-end with Lua scripting, vintagesimulator.com) which I need to stay focused on for now, at least as far as actual programming on the weekends.. I can't actually start coding another thing yet. 

I have been hearing a lot about dat over the last couple of years and that makes me pretty hopeful about p2p data in general.  It sounds as if dat actually works very well.  Performance could be key though.  I think right now this is one of the key challenges for all p2p web projects, since decentralization does not automatically lend itself to fast-loading pages, especially with networks of peers that are relatively small still.

The sidebar with apps thing is a great idea.  The program can be loaded already so it won't slow down page rendering or anything, and you would be in control of it rather than the web page.  Comments is definitely something I want for every web page.

I actually do JavaScript for work for the past several years and just play around with C++/Nim and other stuff on the weekends.  

So I actually think that JavaScript is great and I would never disable it or anything.  I do have problems with performance on mobile web pages on my phone sometimes.  But besides tracking or occasional JavaScript bloat/page lag, really the basis of this idea is something like this: to load a 'normal' web page you pretty much need the entire Chrome /Firefox OS because the API is so huge.. when you do that, they dictate things like extensions and DRM etc. and telemetry to their servers etc.  So it is possible that they could come up with a way to disable ad blocking, or they could create a distribution issue for video content producers with their DRM/new EME thing, etc.  And the core concept of the browser is traditional server-centric web pages.  Also the system makes it easy for governments to censor content or users or block regions.

Practically speaking, maybe with Chromium or WebKit or something you can actually avoid most of those problems.  It seems like there could be some issues at some point though because these massive open source codebases are being maintained by those monopoly corporations which add new features all of the time.  For example, Google and Chrome could add something that breaks Netflix on Chromium or whatever.

With all of the above issues, and the fact that to produce a new implementation of a traditional browser now requires essentially creating a new operating system from scratch, I started to think along the lines of something radically different.  If the developers/users are not going to have a kitchen sink API, what advantage are they going to get?  

The basic idea is that by creating a different paradigm with different constraints, they can get things like privacy, avoid censorship, very fast loading and navigation, and even some fairly useful application capabilities.  And with a new paradigm we don't need to implement an entire operating system since a useful API and featureset can be much smaller if it isn't supposed to be a "real" web page.  Since we don't have all of the web platform features, it might make more sense for people who already dislike JavaScript -- hence the marketing idea to call it "NoScriptWeb".   But still, with web assembly extended to have a simple canvas-like UI API and p2p protocol communication, quite useful highly performant applications could be distributed.

Have you thought about trying to incorporate some type of cryptocurrency wallet into Beaker?  Maybe including MetaMask in the package or something.  Personally I think that the p2p web can be much more useful with integrated decentralized payments.

Do you know of anyone who has tried to build a dat client in C or C++?

**Comment by pfrazee** [comment]
> The performance is also probably slightly better than the other p2p web stuff, but still honestly not as good as the normal web. Maybe some type of pre-caching of links could help? It seems like any pause for dat to load a hash or anything could turn into a big sticking issue with adoption. Because people are impatient.

This is absolutely true. One thing we foresee is "supernodes" which are essentially large caches of dats which the user maintains a continuous connection with. These will enable users to fetch data without hitting the peer-discovery process, and even has the potential to improve loading over the traditional web because it leverages a pre-existing connection.

>  I have a job during the week and another side project (3D libretro front-end with Lua scripting, vintagesimulator.com) 

Neat!

> I have been hearing a lot about dat over the last couple of years and that makes me pretty hopeful about p2p data in general. It sounds as if dat actually works very well. Performance could be key though.

So far so good. We're about to deploy a new version that improves performance, scaling, and reliability. I'm fairly optimistic.

> Practically speaking, maybe with Chromium or WebKit or something you can actually avoid most of those problems. It seems like there could be some issues at some point though because these massive open source codebases are being maintained by those monopoly corporations which add new features all of the time. For example, Google and Chrome could add something that breaks Netflix on Chromium or whatever.

Yes, I think there's a very fair argument to be made that the size of the Web runtime has led to a defacto corporate governance of the Web. Because Chromium is open-source, and because Mozilla is still active, I think it's a situation that can continue for a few more years, but if either org were to wind down, I would be concerned. 

My pet theory that the current system might become untenable around the time that AR/VR become the more dominant computing platforms, and the need to create something better-fitting to those systems will enable a pivot to a fresh approach. 

> The basic idea is that by creating a different paradigm with different constraints, they can get things like privacy, avoid censorship, very fast loading and navigation, and even some fairly useful application capabilities. And with a new paradigm we don't need to implement an entire operating system since a useful API and featureset can be much smaller if it isn't supposed to be a "real" web page. Since we don't have all of the web platform features, it might make more sense for people who already dislike JavaScript -- hence the marketing idea to call it "NoScriptWeb". But still, with web assembly extended to have a simple canvas-like UI API and p2p protocol communication, quite useful highly performant applications could be distributed.

I agree on all those points except the point about the UI - I think you're ultimately going to need a sophisticated GUI runtime because GUIs are just difficult to build. If you strip the GUI tools out of the platform, then web applications will end up building their own and shipping them on the page.

That can still lead to interesting advantages. With proper caching, popular GUI libraries will essentially be pre-downloaded. Then you're in a "micro kernel" situation which, again, helps devolve governance out of the browser, so that's a win. Then, of course, pages that don't need GUIs won't ship them.

With or without a change to the UI layer of the browser, you can still move to a UI-free document model on the Web. Imagine for instance that the dominant document-type on the Web was JSON instead of HTML. When the user browsed to a JSON page, the browser could dynamically load a pre-installed viewer application according to identifying types in the JSON. Now you've gotten much quicker load-times and much more end-user choice over the experience.

> Have you thought about trying to incorporate some type of cryptocurrency wallet into Beaker? Maybe including MetaMask in the package or something. Personally I think that the p2p web can be much more useful with integrated decentralized payments.

We talk about this a lot. None of the crypto-currencies have really impressed us enough to use them, but we very much want a payments layer. One thing we discuss is whether we can find a way to make traditional payment services work without tying the Web platform to a specific provider (ie Paypal). There are signs that traditional payments infrastructure is starting to move in that direction (apparently Visa is exploring a kind of global ID system which they don't own) so I'm optimistic in the long run. We'll see.

> Do you know of anyone who has tried to build a dat client in C or C++?

One of our good friends is working on one, https://github.com/datcxx
